waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -0,0 +1,483 @@
+  (log/info "Creating service" service-id "with description" descriptor)
+  (let [{:strs [run-as-user] :as service-description} (:service-description descriptor)
+        spec-json (service-spec service-id service-description service-id->password-fn)
+        request-url (str api-server-url
+                         "/apis/extensions/v1beta1/namespaces/"
+                         (service-description->namespace service-description)
+                         "/replicasets")
+        response-json (api-request http-client request-url
+                                     :request-method :post
+                                     :body (as-json spec-json))]
+    (replicaset->Service response-json)))
+
+(defn- delete-service
+  [api-server-url http-client service]
+  (when-not service
+    (ss/throw+ {:status 404, :message "Service not found"}))
+  (let [replicaset-url (str api-server-url
+                            "/apis/extensions/v1beta1/namespaces/"
+                            (:namespace service)
+                            "/replicasets/"
+                            (:app-name service))]
+    ; FIXME - catch and handle exceptions
+    (patch-object-json replicaset-url http-client
+                       [{:op :replace, :path "/metadata/annotations/waiter~1app-status", :value "killed"}
+                        {:op :replace, :path "/spec/replicas", :value 0}])
+    (doseq [pod (get-replicaset-pods api-server-url http-client service)
+            :let [pod-url (->> pod :metadata :selfLink (str api-server-url))]]
+      (api-request http-client pod-url :request-method :delete))
+    (api-request http-client replicaset-url :request-method :delete)
+    {:result :deleted
+     :message (str "K8s deleted ReplicaSet " (:app-name service))}))
+
+(defn- service-id->service
+  [{:keys [api-server-url http-client service-id->service-description-fn]} service-id]
+  (let [service-ns (-> service-id service-id->service-description-fn service-description->namespace)
+        replicasets (->> (str api-server-url
+                              "/apis/extensions/v1beta1/namespaces/"
+                              service-ns
+                              "/replicasets?labelSelector=managed-by=waiter,app="
+                              (service-id->app-name service-id))
+                         (api-request http-client)
+                         :items)]
+    (when (seq replicasets)
+      (-> replicasets first replicaset->Service))))
+
+; The KubernetesScheduler
+(defrecord KubernetesScheduler [api-server-url http-client
+                                service-id->service-description-fn]
+  scheduler/ServiceScheduler
+
+  (get-apps->instances [_]
+    (let [apps (get-services api-server-url http-client)
+          _ (log/info "apps =" apps)
+          app->app+instances (partial service+instances api-server-url http-client)
+          app->instances (into {} (mapv app->app+instances apps))
```

Comment:
we could use `pc/map-from-keys` and then `app->app+instances` could just be `app->instances`

Reply:
Yes, that makes this much simpler.